services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java

```diff
@@ -153,6 +155,12 @@ private Response process(MultivaluedMap<String, String> params) {
             return errorResponse;
         }
 
+        try {
+            ClientPolicyManager.triggerOnAuthorizationRequest(parsedResponseType, request, redirectUri, session);
```

Comment:
I see, Instead of calling `ClientPolicyManager` directly, calling `KeycloakSession.clientPolicy()` .

Reply:
I've tried to add `clientPolicy()` method on `KeycloakSession` and `DefaultKecloakSession` but have a problem. `KeycloakSession` belongs to keycloak-server-spi while `ClientPolicyManager` belongs to keycloak-services. The latter is not visible from the former. Considering signatures of methods in `ClientPolicyManager`, it seems to be difficult to implement `ClientPolicyManager` interface in keycloak-server-spi and `DefaultClientPolicyManager implements ClientPolicyManager` class in keycloak-services. Is it inevitably necessary to do so instead of the current `ClientPolicyManager`'s method call directly?

Reply:
I'm not quite sure I'm following. Why would it be problematic to have a ClientPolicyManager interface in keycloak-server-spi, with the DefaultClientPolicyManager in services?

Reply:
I've treated this matter as follows :

- Crate `ClientPolicyContext` in keycloak-server-spi interface only have a method for returning an event.
- `ClientPolicyManager` interface in keycloak-server-spi receives this `ClientPolicyContext` as an argument.
- `KeycloakSession` interface in keycloak-server-spi has the method to get `ClientPolicyManager`.
- An implementation class of `ClientPolicyContext` in keycloak-services implements actual methods depending on an event defined in `ClientPolicyEvent`.